pkg/logcli/query/query.go

```diff
@@ -156,9 +173,93 @@ func (q *Query) DoQuery(c client.Client, out output.LogOutput, statistics bool)
 				// fudge the timestamp forward in time to make sure to get the last entry from this batch in the next query
 				end = lastEntry[0].Timestamp.Add(1 * time.Nanosecond)
 			}
+		}
+	}
+
+	if partFile != nil {
+		if err := partFile.Complete(); err != nil {
+			log.Fatalln(err)
+		}
+	}
+}
+
+func (q *Query) outputFilename() string {
+	return fmt.Sprintf("%s-%d-%d.part", q.PartFilePrefix, q.Start.Unix(), q.End.Unix())
+}
+
+// createPartFile returns a PartFile if the PartFilePrefix is set.
+// The bool value shows if the part file already exists, and this range should be skipped.
+func (q *Query) createPartFile() (*PartFile, bool) {
+	if q.PartFilePrefix == "" {
+		return nil, false
+	}
+
+	partFile := NewPartFile(q.outputFilename())
+
+	if !q.OverwriteCompleted {
+		// If we already have the completed file, no need to download it again.
+		// The user can delete the files if they want to download parts again.
+		exists, err := partFile.Exists()
+		if err != nil {
+			log.Fatalf("Query failed: %s\n", err)
+		}
+		if exists {
+			log.Printf("Skip range: %s - %s: already downloaded\n", q.Start, q.End)
+			return nil, true
+		}
+	}
+
+	if err := partFile.CreateTemp(); err != nil {
+		log.Fatalf("Query failed: %s\n", err)
+	}
+
+	return partFile, false
+}
+
+func (q *Query) DoQueryParallel(c client.Client, out output.LogOutput, statistics bool) {
+	wg := sync.WaitGroup{}
+	jobs := make(chan Query)
+
+	// Start workers
+	for w := 0; w < q.ParallelMaxWorkers; w++ {
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+			for rq := range jobs {
+				rq.DoQuery(c, out, statistics)
+			}
+		}()
+	}
+
+	start := q.Start
+	end := minTime(start.Add(q.ParallelDuration), q.End)
+
+	// Queue up jobs
+	for {
+		rq := *q
+		rq.Start = start
+		rq.End = end
+
+		jobs <- rq
 
+		if end.Equal(q.End) {
+			break
 		}
+
+		start = end
```

Comment:
Checkout the comment in https://github.com/grafana/loki/blob/cf23194d05d33105958f9e1ef70da3ed9717a81c/pkg/logcli/query/query.go#L142 I think you need to specify a limit.